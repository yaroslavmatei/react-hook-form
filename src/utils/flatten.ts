import { FieldValues } from '../types';

export const flatten = (obj: FieldValues) => {
  const output: FieldValues = {};

  for (const key of Object.keys(obj)) {
    if (typeof obj[key] === 'object') {
      const nested = flatten(obj[key]);

      for (const nestedKey of Object.keys(nested)) {
        output[`${key}.${nestedKey}`] = nested[nestedKey];
      }
    } else {
      output[key] = obj[key];
    }
  }

  return output;
};
