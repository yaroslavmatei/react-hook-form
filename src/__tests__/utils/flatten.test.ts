import { flatten } from '../../utils/flatten';

describe('flatten', () => {
  it('should flatten form values into flat form data', () => {
    expect(
      flatten({
        hey: 'test',
        array: [
          {
            test: '1',
            test2: '2',
          },
        ],
        test: {
          nested: {
            test: 'bill',
          },
        },
      }),
    ).toMatchSnapshot();
  });
});
