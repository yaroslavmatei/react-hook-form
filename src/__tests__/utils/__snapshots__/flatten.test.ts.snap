// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flatten should flatten form values into flat form data 1`] = `
{
  "array.0.test": "1",
  "array.0.test2": "2",
  "hey": "test",
  "test.nested.test": "bill",
}
`;
